becquerel/core/calibration.py

```diff
@@ -0,0 +1,787 @@
+            i.e., "p[j]" is the jth parameter, and all parameters must be
+            explicitly indexed in the expression. Can be a single-line formula
+            like "p[0] + p[1] * x" or a code block.
+        params : array_like
+            List of floating point parameters for the calibration function
+        attrs : dict
+            Other information to be stored with the calibration.
+        """
+        self.expression = expression
+        self.params = params
+        self.attrs = attrs
+        self.set_points()
+
+    def __str__(self):
+        """A string version of the calibration."""
+        result = ""
+        result += "expression:\n"
+        lines = str(self.expression).split("\n")
+        for line in lines:
+            result += " " * 4 + line + "\n"
+        result += "params:\n"
+        result += " " * 4 + str(self.params) + "\n"
+        if len(self.points_x) > 0:
+            result += "calibration points (x):\n"
+            result += " " * 4 + str(self.points_x) + "\n"
+            result += "calibration points (y):\n"
+            result += " " * 4 + str(self.points_y) + "\n"
+        if len(self.attrs.keys()) > 0:
+            result += "other attributes:\n"
+            result += " " * 4 + str(self.attrs)
+        return result
+
+    def __repr__(self):
+        """A string representation of the calibration."""
+        result = "Calibration("
+        result += repr(self.expression) + ", "
+        result += repr(self.params)
+        if len(self.attrs) > 0:
+            for key in self.attrs:
+                result += f", {key}={repr(self.attrs[key])}"
+        result += ")"
+        return result
+
+    @staticmethod
+    def eval_expression(expression, params, x):
+        """Evaluate the expression at x.
+
+        Parameters
+        ----------
+        expression : string
+            The expression that defines the calibration function.
+        params : array_like
+            List of floating point parameters for the calibration function
+        x : float or array_like
+            The argument at which to evaluate the expression.
+
+        Returns
+        -------
+        y : float or array_like
+            Result of evaluating the expression for x.
+        """
+        x = np.asarray(x)
+        if not np.all(x >= 0):
+            raise CalibrationError(f"x must be >= 0: {x}")
+        safe_eval.symtable["p"] = params
+        safe_eval.symtable["x"] = x
+        y = safe_eval(expression)
+        if len(safe_eval.error) > 0:
+            raise CalibrationError(
+                f"asteval failed with errors:\n"
+                + "\n".join(str(err.get_error()) for err in safe_eval.error)
+            )
+        if not np.all(np.isreal(y)):
+            raise CalibrationError(
+                f"Function evaluation resulted in complex values: {y}"
+            )
+        # clip values of y
+        y = np.clip(y, 0, CLIP_MAX)
```

Comment:
@markbandstra we should remove this as well. I have a linear calibration with a negative intercept, so the first few bin edges in keV are negative, and all get clipped to 0. Then applying the calibration causes `ValueError: Bin edge energies must be strictly increasing`.

Reply:
Probably means we need both an x- and y-domain specified then.

Reply:
Should be resolved now with the addition of `domain` and `range` to `Calibration`.